### PR TITLE
AppendToSlice - Issue #43

### LINKED ISSUE #43: ToSlice dose not re-empty the dist slice
Codes:

```go
func main() {
	l := []string{"foo", "bar", "baz"}
	linq.From(l).ToSlice(&l)
	fmt.Println(l)
}
```

Result:

```
[foo bar baz foo bar baz]
```

go-linq version:

```
- name: github.com/ahmetalpbalkan/go-linq
  version: 8985ec99e11a8bff7eb19dd0a0b2187770dab23a
```

I don't know whether this is a bug or a feature, but if this is a feature, I think it's better to have this behavior documented on `ToSlice`'s API doc.

### PR DESCRIPTION
**Pull request by cleitonmarx**
Closes #43

### CODE REVIEW

**Comment by ahmetalpbalkan** [comment]
We had #45 to fix this already. I am not sure we should be replacing underlying slices.

**Comment by kalaninja** [comment]
I suppose that replacing slices may result in an unexpected behaviour.
And look at the signature of the method. It expects an existing slice, so it seems really strange to me that someone needs to allocate a slice to call this method and get a new slice.

**Comment by cleitonmarx** [comment]
@ahmetalpbalkan  As a ".NET LINQ-like" library, I think that the behavior of ToSlice should be the same as the original LINQ ToList. In ToList implementation the return is always a [new instance](https://github.com/dotnet/corefx/blob/e6b6ba11c0b7541fb7dc8eac769bf541f5ece3c3/src/System.Linq/src/System/Linq/ToCollection.cs#L30).
A good use case is ordering a slice using linq (example found [here](https://github.com/dotnet/corefx/blob/77f844f3f05bb85d9f6d6b9ae1c1a7d3cef327aa/src/System.Data.SqlClient/tests/Tools/TDS/TDS/PreLogin/TDSPreLoginToken.cs#L161)):

using C#:
```csharp
options = options.OrderBy(o => o.Position).ToList();
```

using Go:
```go
From(options).OrderBy(func(o interface{}) interface{}{
  return o.(Option).Position
}).ToSlice(&options)
```

Currently, the code in Go will not have the same result of the C# code. My Pull Request tries to fix that.

**Comment by kalaninja** [comment]
@cleitonmarx No arguing, you are absolutely right about the .net implementation. But, Go is a different story. Due to strong typing and generics `ToList()` in C# can allocate memory [efficiently](https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Collections/Generic/List.cs#L78). The thing we cannot achieve in Go.
So, the thing that I offer is to rename current method `ToSlice` to `AppendToSlice`, and introduce a new `ToSlice` that works with fully prepared slice. So, if we call it like this:
```go
result := make([]Option, 10)
From(options).OrderBy(func(o interface{}) interface{}{
  return o.(Option).Position
}).ToSlice(&result)
```
`ToSlice` will iterate over the query and put each item to a corresponding index in the slice, so the first item goes to result[0], second to result[1], etc.

@ahmetalpbalkan What do you think?

**Comment by aisk** [comment]
> ToSlice will iterate over the query and put each item to a corresponding index in the slice, so the first item goes to result[0], second to result[1], etc.

Hi have a question, if we have a `Where` call before `ToSlice`, the final result count will be less than the origin. And I think this behavior is surprised to most people.

**Comment by kalaninja** [comment]
@aisk The idea is that if you know what result to expect, you allocate the exact amount of memory that you need and call ToSlice()
```go
result := make([]int, 3)
From([]int{1,2,3,4,5}).Where(func(i interface{}) bool {
			return i.(int) >= 3
		}).ToSlice(&result)
```
if you don't, you append
```go
result := []bool{}
From(data).Where(func(i interface{}) bool {
			return i.(bool)
		}).AppendToSlice(&result)
```

For @cleitonmarx example with sorting
```go
From(options).OrderBy(func(o interface{}) interface{}{
  return o.(Option).Position
}).ToSlice(&options)
```
will make an in-place ordering without allocating a new slice inside `ToSlice`

**Comment by aisk** [comment]
@kalaninja  Ok I know. But I think this method name `ToSlice` is misunderstanding to some people(like me). I think `ToSlice` is a generic and common name that will do all the magic for me like what I'm expecting, like allocate memory fast or replace current slice's elements and adjust it to excepted size.

So I think if we should keep the two `unexpected` method (they have the behavior that is not so strait to most people, because the performance case), I think we should use an also `unexpected` method name, like `ReplaceToSlice` and `AppenedToSlice`, and leave the `ToSlice` method name until go have some way to resolve the performance case.

**Comment by ahmetalpbalkan** [comment]
I think we should keep `ToSlice()` the same (that keeps appending) with some heavy documentation that it doesn't replace the slice. There might be people using it since pre-v1.0.

the `AppendToSlice` sounds a bit long to me and doesn't follow the `ToXxx` pattern, how about:
- `ToSlice(&v)`: appends
- `ToNewSlice(): []interface{}` returns a new slice

**Comment by cleitonmarx** [comment]
@kalaninja, I think that will be hard, is some scenarios, know the size of the resulting slice in ```ToSlice()``` method. One question, if append a slice is so bad, why are we using the same pattern for ```ToSlice()``` currently? 
Another proposal to implement it will be like this:
```go
func (q Query) ToSlice(result interface{}) {
	res := reflect.ValueOf(result)
	typeResult := res.Type().Elem()
	count := q.Count()
	slice := reflect.MakeSlice(typeResult, count, count)
	count = 0

	next := q.Iterate()
	for item, ok := next(); ok; item, ok = next() {
		slice.Index(count).Set(reflect.ValueOf(item))
		count++
	}

	res.Elem().Set(slice)
}
```
But in this case, I think is not a good idea Iterate over the Query twice.

@ahmetalpbalkan, I think this pattern ```ToNewSlice(): []interface{}``` is an antipattern, because I will always need to convert the result to a typed slice. I think the pattern used in V2 with ```ToSlice()``` solved this issue and we need to keep it.

**Comment by cleitonmarx** [comment]
@kalaninja Actually, calling ```MakeSlice()``` using the ```Query.Count()``` improved performance and memory allocation, check it out:
```
BenchmarkToSlice_UsingMakeSlice-4         	      20	  96989273 ns/op	24003928 B/op	 2000010 allocs/op
BenchmarkToSlice_UsingAppend-4            	      10	 155321097 ns/op	81418236 B/op	 2000047 allocs/op
BenchmarkAppendToSlice_UsingMakeSlice-4   	      20	  95919171 ns/op	24003841 B/op	 2000010 allocs/op

```

**Comment by kalaninja** [comment]
@ahmetalpbalkan I like your point about naming pattern. Using conventions makes you code more understandable and easier to use.

> I think that will be hard, in some scenarios, know the size of the resulting slice in ToSlice() method

@cleitonmarx what I offer is that you use `ToSlice` only if you know the size, and if you don't you allocate an empty slice and use `AppendToSlice` (or whatever its name might be)

> One question, if append a slice is so bad for memory allocation, why are we using the same pattern for ToSlice() currently?

It's might fault, I didn't think about it while writing the code.

>  I think is not a good idea Iterate over the Query twice using Count and q.Iterate().

Absolutely agree. Iterating a query two times is something we cannot afford since no iterator is guaranteed to be idempotent.

> calling MakeSlice() using the Query.Count() improved performance and memory allocation

I assume that you have a very simple query. So that the second iteration is less consuming than slice appending. Add some sorting and the situation will change.

**Comment by cleitonmarx** [comment]
> I assume that you have a very simple query. So that the second iteration is less consuming than slice appending. Add some sorting and the situation will change.

Yeah, You are totally right. I could see your point putting more complexity on the benchmark.

**Comment by ahmetalpbalkan** [comment]
The runtime performance assumption in `ToSlice()` is basically, if I have input of, say, 100,000 elements and I know that my query will match to ~2,000 results, I can allocate a slice with cap=2000 and pass that to `ToSlice()`. If I get more results, the `append` will double the capacity and copy N=cap elements over to the new slice. However if you give it a small slice (e.g. cap=1) and have N=2000 results, it'll double (like 2, 4, 8, 16, ...) and will take lg(N,2) steps to reach to cap=2048.

**Comment by kalaninja** [comment]
@ahmetalpbalkan That is actually a good idea. I can reimplement `ToSlice` with this logic. So that you allocate a slice of 2000 elements. Call `ToSlice` and the method uses this memory, appends items if initial capacity is not enough and returns subslice if initial capacity is excessive. How about this?

**Comment by ahmetalpbalkan** [comment]
@kalaninja isn't that already what happens today, if I give you a slice of len=0 cap=100 (`make([]T,0,100)` ?

**Comment by cleitonmarx** [comment]
@ahmetalpbalkan @kalaninja I pushed the optimization described by both of you using the lg(N,2) algorithm in memory allocation. Please review it.

File: result_test.go [hunk 1]
```diff
@@ -409,11 +409,13 @@ func TestToMapBy(t *testing.T) {
 
 func TestToSlice(t *testing.T) {
```
Comment (ahmetalpbalkan) [inline result_test.go]: can we have tests for edge cases:

1. output is nil slice.
2. cap(out)>len(result): we get the same slice, just resized.
3. cap(out)==len(result): we get the same slice, just resized.
4. cap(out)<len(result): we get a new slice with len(out)=len(result) and cap(out)==??
5. len(out)>0 and appending/growing doesn't lose the original elements in out.

Comment (cleitonmarx) [inline result_test.go]: > len(out)>0 and appending/growing doesn't lose the original elements in out.
@ahmetalpbalkan the behavior in this scenario will be: copy the result reusing the len(out) and appending if len(result) > len(out). In this case output slice will be overwritten(and appended if necessary) with the result elements. If len(out) > len(result) the elements of result will be copied to the output and resized to len(result). What do you think?

Comment (ahmetalpbalkan) [inline result_test.go]: Hmm the godoc for `ToSlice` doesn't really tell me that it starts overwriting the given slice from index=0. Is that the behavior? Perhaps we should rewrite `ToSlice` godoc?

File: result_test.go [hunk 2]
```diff
@@ -408,12 +408,31 @@ func TestToMapBy(t *testing.T) {
...
+	tests := []struct {
+		input        []int
+		output       []int
+		want         []int
+		wantedOutCap int
+	}{
+		//output is nil slice
+		{[]int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, []int{}, []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 16},
+		//cap(out)>len(result): we get the same slice, just resized.
+		{[]int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, make([]int, 0, 11), []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 11},
```
Comment (ahmetalpbalkan) [inline result_test.go]: @cleitonmarx you're not testing the `we get the same slice` part. :) that'd be a pointer comparison.

File: result.go [hunk 3]
```diff
@@ -381,17 +381,42 @@ func (q Query) ToMapBy(
 	res.Elem().Set(m)
 }
 
-// ToSlice iterates over a collection and populates the result slice with elements.
+// ToSlice iterates over a collection, copy to result slice the collection elements,
```
Comment (ahmetalpbalkan) [inline result.go]: Let's fix this godoc. It's confusing to me, as it says:

1. "iterates over a collection" ––do we mean "slice"? what are some other possible collections here?
2. "copy to result slice the collection elements" ––sounds like a grammatical mistake here?
3. "appends items if initial capacity is not enough" ––probably should sound more like: _"If it has sufficient capacity, the destination is resliced to accommodate the new elements. If it does not, a new underlying array will be allocated."_ [[source](https://golang.org/pkg/builtin/#append)]
4. we should still indicate that the slice is resliced to reflect the size of the result.

**Comment by ahmetalpbalkan** [comment]
BTW I'll edit this PR to change the base branch to `v3-dev` as well. Since it's a behavior change, we should bring this in as part of v3.

File: result.go (hunk 3 above)
Comment (cleitonmarx) [inline result.go]: > "iterates over a collection" ––do we mean "slice"? what are some other possible collections here?

@ahmetalpbalkan I don't know, I'm just following the current documentation terminology like [here](../master/result.go#L333) , [here](../master/result.go#L345)  and [here](../master/result.go#L360) 

> "copy to result slice the collection elements" ––sounds like a grammatical mistake here?

Sorry man, English is not my first language. Will be amazing have more help with documentation.

Comment (ahmetalpbalkan) [inline result.go]: @cleitonmarx no problem! None of us are native English speakers here. :) If you could check the box for "Allow edits from maintainers" on the right menu, I can help you with this commit.

File: result_test.go [hunk 4]
```diff
@@ -409,30 +409,42 @@ func TestToMapBy(t *testing.T) {
...
+		modifiedOutputValue := reflect.ValueOf(test.output)
 		if !reflect.DeepEqual(test.output, test.want) {
 			t.Errorf("From(%v).ToSlice()=%v expected %v", test.input, test.output, test.want)
 		}
-		if cap(test.output) != test.wantedOutCap {
-			t.Errorf("cap(output)=%d expected %d", cap(test.output), test.wantedOutCap)
+		if cap(test.output) != test.wantedOutputCap {
+			t.Errorf("cap(output)=%d expected %d", cap(test.output), test.wantedOutputCap)
+		}
+		isNewSlice := initialOutputValue.Pointer() != modifiedOutputValue.Pointer()
```
Comment (ahmetalpbalkan) [inline result_test.go]: you can just do this to get uintptr: `(*reflect.SliceHeader)(unsafe.Pointer(&v)).Data`

**Comment by ahmetalpbalkan** [comment]
@cleitonmarx just pushed 75d6687 reorganizing the test a little bit. While I'm at it, I updated some test case comments and added new test cases. I simplified the pointer equality check as well. PTAL

File: result.go (hunk 3 above)
Comment (ahmetalpbalkan) [inline result.go]: @cleitonmarx I now addreesed this. Please take a look at the new `ToSlice()` doc.
cc: @kalaninja  as well.

Comment (cleitonmarx) [inline result.go]: Looks great for me @ahmetalpbalkan 👍 , thanks for helping me with documentation.

Comment (ahmetalpbalkan) [inline result.go]: lol forgot to push the commit here. see e354c24.

**Comment by cleitonmarx** [comment]
@ahmetalpbalkan I liked the new ToSlice tests scenarios. Looking forward to merging this changes and start to upgrade go-linq v3 in my application. :tada:

**Comment by ahmetalpbalkan** [comment]
Changed base branch to `v3-dev` and rebased.

File: result.go [hunk 5]
```diff
@@ -546,17 +546,46 @@ func (q Query) ToMapByT(result interface{}, keySelectorFn interface{}, valueSele
...
+func (q Query) ToSlice(v interface{}) {
+	res := reflect.ValueOf(v)
 	slice := reflect.Indirect(res)
 	next := q.Iterate()
-
+	index := 0
 	for item, ok := next(); ok; item, ok = next() {
-		slice = reflect.Append(slice, reflect.ValueOf(item))
+		if index >= slice.Len() {
+			slice = grow(slice, 1)
```
Comment (ahmetalpbalkan) [inline result.go]: @cleitonmarx Sorry it's taking too many iterations but I think I found a critical perf issue:

 I know that we don't know the result side in advance, but it looks like this code is not super optimal.

1. `slice.Len()`  (which contains a type switch == slow) here
2. `grow(slice,1)`

every time we add an element (`N` times). I think this algorithm is inefficient. How about the following one:

- we read the cap of slice `s`, and reslice the `s` to its full capacity (`cap(s)`). so now `len(s)==cap(s)`.
- store this `len:=s.Len()`  into a variable.
- we go into the for loop and to add elements
    - if `index >= len`, we grow the slice by a factor of 2 (meaning, double the slice)
        - if we do this, then **reslice** the `s` to the new `s.Cap()` and update `len = s.Cap()` as well
    - else
        - just set the element at `index`
- before returning, **reslice** s to `0:index`

This way at the end, we'll still keep doubling the cap oft he given slice --however, we're not increasing len one at a time. in this algorithm, it will be always `len(s) == cap(s)` before the last step which resizes. By doing that:
- you actually call `grow()` only a few times (logarithmic)
- no more `N` calls to `s.Len()` as you store `len`/`cap` in variables as you grow
- Perhaps `grow()` can also return the new capacity, this way you don't actually ever need to  call `.Cap()` in the algorithm above.

Can we implement `ToSlice()` like this?

That said, I'd ❤️  to see how this current algorithm vs the one I proposed above behaves in a benchmark.

Comment (ahmetalpbalkan) [inline result.go]: Actually, we can do this in another follow-up PR. Let me know if you want to keep working on the current PR to implement the proposed algorithm, or if you want this to just get merged and you can open a PR later.

Comment (cleitonmarx) [inline result.go]: @ahmetalpbalkan Let's do this!
Here my first try:

```go

Comment (cleitonmarx) [inline result.go]: Ok, It makes sense to me. I will try to send a commit with this changes tonight.

Comment (ahmetalpbalkan) [inline result.go]: @cleitonmarx I just pushed the changes I suggested above (1d3f5b0). (you can easily tell I'm enjoying free time between jobs 🤣) It passes the tests, let me know if it looks good to you.

**Comment by ahmetalpbalkan** [comment]
I'll simplify the commit history and merge.
